### PR TITLE
discovering broker and proxying requests to broker

### PR DESCRIPTION
**Pull request by himanshug**

### CODE REVIEW

File: docs/content/configuration/index.md [hunk 1]
```diff
@@ -257,6 +257,14 @@ This config is used to find the [Indexing Service](../design/indexing-service.ht
 |--------|-----------|-------|
 |`druid.selectors.indexing.serviceName`|The druid.service name of the indexing service Overlord node. To start the Overlord with a different name, set it with this property. |druid/overlord|
 
+### Broker Discovery
+
+This config is used to find the [Brokers](../design/broker.html) using Curator service discovery.
+
+|Property|Description|Default|
+|--------|-----------|-------|
+|`druid.selectors.broker.serviceName`|The druid.service name of the broker node. |druid/broker|
```
Comment (gianm) [inline docs/content/configuration/index.md]: Our tutorials and examples use "broker" for the broker name, so maybe we should use that as the default here.

Comment (gianm) [inline docs/content/configuration/index.md]: Not everyone makes queries to brokers directly, some people use routers. So if the intent is to proxy to whatever the normal query thinger is, maybe this should be something like ```druid.selectors.query.serviceName```? Or something?

Comment (gianm) [inline docs/content/configuration/index.md]: There is also not necessarily going to be just one kind of broker per cluster. Some people have different brokers they use for different use cases.

Comment (himanshug) [inline docs/content/configuration/index.md]: 1) choosing default being druid/broker is inline with druid.selectors.indexing.serviceName default druid/overlord , I will change example config files to set it to "broker"
2) will change property name to `druid.selectors.query.serviceName`
3) yes, but for the use case of ui reaching brokers via coordinator for metadata stuff, users should be able to set a service name that is published by a subset of all brokers.

Comment (himanshug) [inline docs/content/configuration/index.md]: 2) actually keeping `druid.selectors.broker.serviceName` is OK as this is intended for coordinator to proxy requests to brokers (originally coming from something like UI). Are there use cases where we would want to proxy requests from coordinator to router instead?

**Comment by xvrl** [comment]
@himanshug can you provide a bit of context on what the objective is here?

If the objective is to provide a single entrypoint for a UI, I feel we should have a broader discussion on defining one entrypoint for everything, instead of adding more configuration parameters and proxies between nodes.

**Comment by himanshug** [comment]
@xvrl the objective is 
(1) proxying is for ppl to write UI  that can query broker endopints regarding metadata via broker, similar setup exists for overlord for the same reason.
(2) broker awareness is also there for coordinator to be able to send segment metadata query to broker so that it can get the list of aggregators for various segments to put them in "merge/reindex" task.

**Comment by xvrl** [comment]
@himanshug 
As far as UI interaction goes, most of it already goes through the broker. For that I feel it would make more sense if the proxying happened on the broker / router side. The main reason we proxy overlord from coordinator is because we plan to merge those anyways.

Proxying from the broker side, would also be more natural for people using router as entry-point or multiple brokers, since everything will flow through from those endpoints. We could have a generic implementation that we could reuse interchangeably either at the router or broker level.

As far as segment metadata is concerned, is the broker really necessary? The coordinator can already query segment metadata from the historical nodes directly, it shouldn't need the broker for that.

**Comment by himanshug** [comment]
@xvrl 
1) for the UI, a colleague of mine want to make updates to coordinator console that needs some information from broker endpoints, not sure how to enable that without proxying on coordinator.
2) for coordinator to directly send query to historicals, it has to build the pipeline of query runners , not sure if that is better or worse.
anyways, we can surely discuss the use cases and see what would be the best way to enable them.

**Comment by drcrallen** [comment]
@himanshug / @fjy can the desire for this to be in 0.9.0 be resolved? If it needs to be in 0.9.0 it needs rebased

**Comment by xvrl** [comment]
@himanshug @fjy maybe it would be worth using whichever abstraction we develop as part of #2040 for internal discovery.

**Comment by xvrl** [comment]
@himanshug I think can get behind this PR if the intention is to offer a simple entry-point to make queries. However, I think we should consider re-using the router servlet to implement request forwarding instead of implementing yet another proxy servlet. It shouldn't be too hard to abstract the router-specific pieces to make it re-usable. Happy to help you out if you're interested.

**Comment by himanshug** [comment]
@xvrl this PR should totally use the new proposed abstractions for discovery.
main use case that drove this PR was for re-indexing proposed in #1998 to be able to get aggregators from segment metadata query and coordinator would need to know of brokers(or router) to be able to send the query.
since all realtime ingestion work is being postponed to 0.9.1, so we can sleep on it for now.

**Comment by fjy** [comment]
@himanshug this has merge conflicts

**Comment by himanshug** [comment]
closing this for now, will revisit if/when needed.

**Pull request closed without merging**